### PR TITLE
More shoulds

### PR DESCRIPTION
**Pull request by EddieGarmon**
Sorry for putting all this in one PR. They are all each separate commits.

- Adding  support for ShouldBe(bool?)
- Implicitly casting ShouldlyMessage to string, no more .ToString(), and cleanup call sites
- Consistent ordering in ShouldlyConventions.ShouldHaveCustomMessageOverloads
- Adding IEqualityComparer<T> support for ShouldBe, ShouldNotBe, ShouldContain, ShouldNotContain

### CODE REVIEW

**Comment by jahmai** [comment]
Will this ever get merged?

File: src/Shouldly.Tests/Shouldly.Tests.csproj [hunk 1]
```diff
@@ -9,6 +9,10 @@
     <GenerateAssemblyProductAttribute>false</GenerateAssemblyProductAttribute>
     <DebugType>full</DebugType>
   </PropertyGroup>
+  <PropertyGroup Condition=" '$(TargetFramework)' == 'netcoreapp2.0' ">
```
Comment (SimonCropp) [inline src/Shouldly.Tests/Shouldly.Tests.csproj:12]: why is this required?

File: GitVersionConfig.yaml [hunk 2]
```diff
@@ -1 +1 @@
-next-version: 3.0.0
\ No newline at end of file
+next-version: 3.1.0
```
Comment (SimonCropp) [inline GitVersionConfig.yaml:1]: i dont think this should have changed

File: src/Shouldly.Tests/ShouldBeBoolean/FalseScenario.cs [hunk 3]
```diff
@@ -10,7 +10,7 @@ public void FalseScenarioShouldFail()
         {
             const bool myValue = true;
             Verify.ShouldFail(() =>
-                myValue.ShouldBeFalse("Some additional context"),
+myValue.ShouldBeFalse("Some additional context"),
```
Comment (SimonCropp) [inline src/Shouldly.Tests/ShouldBeBoolean/FalseScenario.cs:13]: why the indent change?

**Comment by SimonCropp** [comment]
closing this one for now. it is too difficult to merge. sorry for this being ignored for so long. please let me know if you need any help re-submitting
